File: src/2101_detonate_the_maximum_bombs.js
function maximumDetonation(bombs) {
    const n = bombs.length;
    const graph = Array.from({ length: n }, () => []);
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (i !== j) {
                const [x1, y1, r1] = bombs[i];
                const [x2, y2] = bombs[j];
                const dst = Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2);
                if (dst <= r1) graph[i].push(j);
            }
        }
    }
    const dfs = (node, vis) => {
        vis[node] = true;
        let count = 1;
        for (const nbh of graph[node]) {
            if (!vis[nbh]) count += dfs(nbh, vis);
        }
        return count;
    };
    let detonated = 0;
    for (let i = 0; i < n; i++) {
        detonated = Math.max(detonated, dfs(i, Array(n).fill(false)));
    }
    return detonated;
}

module.exports = { maximumDetonation };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(maximumDetonation([[2, 1, 3], [6, 1, 4]]), 2);
    assert.deepStrictEqual(maximumDetonation([[1, 1, 5], [10, 10, 5]]), 1);
    assert.deepStrictEqual(maximumDetonation([[1, 2, 3], [2, 3, 1], [3, 4, 2], [4, 5, 3], [5, 6, 4]]), 5);
}

File: src/2101_detonate_the_maximum_bombs.rs
struct Solution;

impl Solution {
    fn detonation_dfs(node: usize, vis: &mut [bool], graph: &[Vec<usize>]) -> i32 {
        vis[node] = true;
        let mut count = 1;

        for &nbh in &graph[node] {
            if !vis[nbh] {
                count += Self::detonation_dfs(nbh, vis, graph);
            }
        }

        count
    }

    pub fn maximum_detonation(bombs: Vec<Vec<i32>>) -> i32 {
        let n = bombs.len();
        let mut graph = vec![Vec::new(); n];

        for i in 0..n {
            for j in 0..n {
                if i != j {
                    let x1 = bombs[i][0];
                    let y1 = bombs[i][1];
                    let r1 = bombs[i][2];
                    let x2 = bombs[j][0];
                    let y2 = bombs[j][1];
                    let dst = f64::from((x1 - x2).pow(2) + (y1 - y2).pow(2)).sqrt();

                    if dst <= f64::from(r1) {
                        graph[i].push(j);
                    }
                }
            }
        }

        let mut detonated = 0;
        for i in 0..n {
            let mut visited = vec![false; n];
            detonated = detonated.max(Self::detonation_dfs(i, &mut visited, &graph));
        }
        detonated
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::maximum_detonation(vec![vec![2, 1, 3], vec![6, 1, 4]]),
            2
        );
        assert_eq!(
            Solution::maximum_detonation(vec![vec![1, 1, 5], vec![10, 10, 5]]),
            1
        );
        assert_eq!(
            Solution::maximum_detonation(vec![
                vec![1, 2, 3],
                vec![2, 3, 1],
                vec![3, 4, 2],
                vec![4, 5, 3],
                vec![5, 6, 4]
            ]),
            5
        );
    }
}
